src/neuroglancer/datasource/computed/frontend.ts

```diff
@@ -0,0 +1,309 @@
+      cancellationToken: CancellationToken) {
+    if (!options.dataSourceProvider) {
+      return Promise.reject('Need a DataSourceProvider');
+    }
+
+    const dataSourceProvider = options.dataSourceProvider!;
+    const [sourceNames, sourceConfigs] = this.parseConfig(config, dataSourceProvider);
+
+    if (sourceNames.length < 2) {
+      return Promise.reject();
+    }
+
+    // The first source name should be the key to a VolumeComputation
+    const computationProvider = ComputedDataSource.computationMap.get(sourceNames[0]);
+    const originUrl = sourceNames[1] + '://' + sourceConfigs[1];
+
+    if (!computationProvider) {
+      return Promise.reject();
+    }
+
+    return this
+        .getOriginVolumes(
+            dataSourceProvider, originUrl, sourceConfigs[1], chunkManager, cancellationToken)
+        .then((volumes) => {
+          const dataSourceParams = this.defaultParams(volumes, originUrl, dataSourceProvider);
+          const computation = computationProvider.makeComputation();
+          return computation.initialize(sourceConfigs[0], volumes.sources, dataSourceParams)
+              .then((computationParams: ComputationParameters) => {
+                computation.initializeCounterpart(chunkManager.rpc!, computationParams);
+                return new ComputedMultiscaleVolumeChunkSource(
+                    dataSourceParams, volumes.sources, computation, chunkManager);
+              });
+        });
+  }
+
+  static computationMap = new Map<string, VolumeComputationFrontendProvider<any>>();
+
+  static registerComputation(
+      key: string, computationProvider: VolumeComputationFrontendProvider<any>) {
+    this.computationMap.set(key, computationProvider);
+  }
+}
+
+const MAX_FETCH_TRIES = 5;
+const FETCH_DELAY = 50;  // ms
+
+function fetchFrontendChunkData(rpc: RPC, x: any) {
+  const source: VolumeChunkSource = rpc.getRef(x['originSourceRef']);
+  const chunkKey: string = x['chunkKey'];
+  const originGridKey: string = x['originGridKey'];
+  const requestorSourceRef: any = x['requestorSourceRef'];
+  if (!x.hasOwnProperty('try')) {
+    x['try'] = 1;
+  }
+  const nTry = x['try'];
+
+  if (source.chunks.has(chunkKey)) {
+    const chunk = <UncompressedVolumeChunk>source.chunks.get(chunkKey);
+    rpc.invoke(
+        RETURN_FRONTEND_CHUNK,
+        {requestorSourceRef, chunkKey, originGridKey, error: undefined, data: chunk.data});
+    return;
+  }
+
+  // When we can't find data on the frontend, it usually means that we're in a
+  // race condition: we started listening just after the origin chunk's data
+  // was sent to the frontend, but this request arrived before the data was
+  // fully transferred. This is a fairly rare occurrence, but it does happen.
+  // Here, we wait for progressively longer periods of time, giving up after
+  // about 1.25s.
+  //
+  // This is simpler than a listener-callback structure, but is not
+  // deterministic.
+
```

Comment:
Messages are always in order, but the issue this is trying to solve is caused by the Chunk.update logic in chunk_manager/frontend.ts, which defers handling of chunk update requests from the backend in order to avoid harming interactivity of the UI thread.  It would make sense to integrate this RETURN_FRONTEND_REQUEST handling with the same Chunk.update queue so that you don't have to retry.